pkg/oc/admin/diagnostics/diagnostics/cluster/app_create/main.go

```diff
@@ -0,0 +1,357 @@
+		{"admission-timeout", "Seconds to wait for the route to be admitted by a router", &d.routeAdmissionTimeout, AppCreateRouteAdmissionTimeoutDefault},
+		{"skip-service-connect", "Do not test connecting to the service", &d.skipServiceConnect, false},
+		{"skip-route-test", "Do not test route at all", &d.skipRouteTest, false},
+		{"skip-route-connect", "Do not test connecting to the route", &d.skipRouteConnect, false},
+		{"http-timeout", "Milliseconds to wait for an HTTP request to the app", &d.httpTimeout, AppCreateHttpTimeoutDefault},
+		{"http-retries", "Number of times to retry an HTTP request to the app", &d.httpRetries, AppCreateHttpRetriesDefault},
+		{"node-selector", "Node selector for where the test app should land", &d.nodeSelector, ""},
+		{"keep-app", "Do not delete the test app when complete", &d.keepApp, false},
+		{"result-dir", "Directory in which to write result details if desired", &d.writeResultDir, ""},
+	}
+}
+
+func getDefaultAppImage() string {
+	template := variable.NewDefaultImageTemplate()
+	return template.ExpandOrDie("deployer")
+}
+
+func (d *AppCreate) Complete(logger *log.Logger) error {
+	// project management
+	d.keepProject = d.keepProject || d.keepApp // don't delete project if keeping app
+	if d.project == "" && d.projectBase == "" {
+		return fmt.Errorf("%s project name cannot be empty", AppCreateName)
+	}
+	if d.project == "" {
+		// generate a project if not specified
+		d.project = names.SimpleNameGenerator.GenerateName(d.projectBase)
+	} else {
+		// when an existing project is specified, deleting it is likely to surprise the user, so don't
+		d.keepProject = true
+	}
+	if errs := kvalidation.IsDNS1123Label(d.project); len(errs) > 0 {
+		return fmt.Errorf("invalid project name '%s' for AppCreate: %v", d.project, errs)
+	}
+	// TODO: also test that route is valid under DNS952
+
+	// app management
+	if d.appName == "" {
+		return fmt.Errorf("%s app name cannot be empty", AppCreateName)
+	}
+	if errs := kvalidation.IsDNS1123Label(d.appName); len(errs) > 0 {
+		return fmt.Errorf("invalid app name '%s' for AppCreate: %v", d.appName, errs)
+	}
+	if err := kvalidation.IsValidPortNum(d.appPort); err != nil {
+		return fmt.Errorf("invalid app port %d for AppCreate: %v", d.appPort, err)
+	}
+	d.label = map[string]string{"app": d.appName}
+	d.labelSelector = fmt.Sprintf("app=%s", d.appName)
+
+	d.skipRouteConnect = d.skipRouteConnect || d.skipRouteTest // don't try to connect to route if skipping route test
+
+	return nil
+}
+
+func (d *AppCreate) CanRun() (bool, error) {
+	if d.SARClient == nil || d.AppsClient == nil || d.KubeClient == nil || d.ProjectClient == nil || d.RoleBindingClient == nil {
+		return false, fmt.Errorf("missing at least one client")
+	}
+	if d.PreventModification {
+		return false, fmt.Errorf("requires modifications: create a project and application")
+	}
+	return util.UserCan(d.SARClient, &authorization.ResourceAttributes{
```

Comment:
Shouldn't that check more than just creating namespace? I mean explicit checks for the ability to create DC, route, service.

Reply:
@soltysh actually this is rather redundant since running this diagnostic at all requires cluster-admin. The client is created and tested [here](https://github.com/openshift/origin/blob/4f709b48f8e52e8c6012bd8b91945f022a437a6a/pkg/oc/admin/diagnostics/cluster.go#L165) and if it lacks access, no cluster diagnostics run. I don't remember offhand if there was a particular reason for similar checks in the other cluster diagnostics (e.g. future-proofing / fail-safe like checking that clients are defined) or if it should just be considered cruft and removed.

Reply:
Ok, thanks for the explanation.